json_rpc/client.nim

```diff
@@ -34,62 +31,38 @@ method call*(client: RpcClient, name: string,
 
 method close*(client: RpcClient) {.base, gcsafe, async.} = discard
 
-template asyncRaise[T](fut: Future[T], errType: typedesc, msg: string) =
-  fut.fail(newException(errType, msg))
-
-macro checkGet(node: JsonNode, fieldName: string,
-               jKind: static[JsonNodeKind]): untyped =
-  let n = genSym(ident = "n") #`node`{`fieldName`}
-  result = quote:
-    let `n` = `node`{`fieldname`}
-    if `n`.isNil or `n`.kind == JNull:
-      raise newException(ValueError,
-        "Message is missing required field \"" & `fieldName` & "\"")
-    if `n`.kind != `jKind`.JsonNodeKind:
-      raise newException(ValueError,
-        "Expected " & $(`jKind`.JsonNodeKind) & ", got " & $`n`.kind)
-  case jKind
-  of JBool: result.add(quote do: `n`.getBool)
-  of JInt: result.add(quote do: `n`.getInt)
-  of JString: result.add(quote do: `n`.getStr)
-  of JFloat: result.add(quote do: `n`.getFloat)
-  of JObject: result.add(quote do: `n`.getObject)
-  else: discard
-
-proc processMessage*(self: RpcClient, line: string) {.gcsafe.} =
+template `or`(a: JsonNode, b: typed): JsonNode =
+  if a == nil: b else: a
+
+proc processMessage*(self: RpcClient, line: string) =
   # Note: this doesn't use any transport code so doesn't need to be
   # differentiated.
   let node = parseJson(line)
 
   if "id" in node:
-    let id = checkGet(node, "id", JInt)
+    let id = node{"id"} or newJNull()
 
     var requestFut: Future[Response]
-    if not self.awaiting.pop(id, requestFut):
-      raise newException(ValueError,
-        "Cannot find message id \"" & $node["id"].getInt & "\"")
+    if not self.awaiting.pop(id.getInt(-1), requestFut):
+      raise newException(ValueError, "Cannot find message id \"" & $id & "\"")
 
-    let version = checkGet(node, "jsonrpc", JString)
+    let version = node{"jsonrpc"}.getStr()
     if version != "2.0":
-      self.awaiting[id].asyncRaise(ValueError,
-        "Unsupported version of JSON, expected 2.0, received \"" & version & "\"")
-
-    let errorNode = node{"error"}
-    if errorNode.isNil or errorNode.kind == JNull:
-      var res = node{"result"}
-      if not res.isNil:
-        requestFut.complete((false, res))
-      else:
-        requestFut.fail(newException(InvalidResponse, "Missing `result` field"))
+      requestFut.fail(newException(ValueError,
+        "Unsupported version of JSON, expected 2.0, received \"" & version & "\""))
```

Comment:
it would take a bunch of code because stdlib `json.$` crashes when given a `nil` node which `{}` returns, didn't seem like it was worth the effort for the relatively fringe problem here